File: client/src/types/SeasonSummary.ts
const Conference = {
    West: 'West',
    East: 'East',
} as const
type Conference = (typeof Conference)[keyof typeof Conference]
export const conferences = Object.values(Conference)

const Division = {
    Northwest: 'Northwest',
    Pacific: 'Pacific',
    Southwest: 'Southwest',
    Atlantic: 'Atlantic',
    Central: 'Central',
    Southeast: 'Southeast',
}
type Division = (typeof Division)[keyof typeof Division]
export const divisions = Object.values(Division)

interface IRegularSeasonTeamStats {
    id: number,
    reb: number,
    eff: number,
    teamAbbreviation: string,
    teamLogo: string,
    conference: Conference,
    conferenceRank: number,
    division: Division,
    divisionRank: number,
    win: number,
    lose: number,
    pts: number,
    ast: number,
    stl: number,
    blk: number,
    fg: number,
    fga: number,
    three: number,
    threea: number,
    ft: number,
    fta: number,
    oreb: number,
    dreb: number,
    to: number,
    pf: number,
    plusminus: number,
    teamId: number,
    season: string,
}

export interface ISeasonSummary {
    season: string,
    regularSeasonTeamsStats: IRegularSeasonTeamStats[]
}

export interface RegularSeasonTeamStats {
    teamAbbreviation: string,
    teamLogo: string,
    conference: Conference,
    conferenceRank: number,
    division: Division,
    divisionRank: number,
    win: number,
    lose: number,
    pct: number,
    gb: number,
}

export class SeasonSummary {
    season: string
    regularSeasonTeamsStats: RegularSeasonTeamStats[]

    constructor(data?: ISeasonSummary) {
        this.season = data?.season ?? ''
        this.regularSeasonTeamsStats = []
        if (data) {
            const westMaxDiff = data.regularSeasonTeamsStats.reduce((max, item) => {
                const diff = item.win - item.lose
                return diff > max && item.conference == Conference.West ? diff : max
            }, -Infinity)
            const eastMaxDiff = data.regularSeasonTeamsStats.reduce((max, item) => {
                const diff = item.win - item.lose
                return diff > max && item.conference == Conference.East ? diff : max
            }, -Infinity)
            for (const regularSeasonTeamStats of data?.regularSeasonTeamsStats) {
                const win = regularSeasonTeamStats.win
                const lose = regularSeasonTeamStats.lose
                const maxDiff = regularSeasonTeamStats.conference == Conference.West ? westMaxDiff : eastMaxDiff
                this.regularSeasonTeamsStats.push({
                    teamAbbreviation: regularSeasonTeamStats.teamAbbreviation,
                    teamLogo: regularSeasonTeamStats.teamLogo,
                    conference: regularSeasonTeamStats.conference,
                    conferenceRank: regularSeasonTeamStats.conferenceRank,
                    division: regularSeasonTeamStats.division,
                    divisionRank: regularSeasonTeamStats.divisionRank,
                    win: win,
                    lose: lose,
                    pct: win + lose > 0 ? Math.round(win / (win + lose) * 10) / 10 : 0,
                    gb: (maxDiff - (win - lose)) / 2
                })
            }
        }
    }
}

File: server/rest_api/models/season_summary.py
from django.db import models

from rest_api.models.game_summary import Team


class SeasonSummary(models.Model):
    season = models.CharField(primary_key=True)
    @property
    def regular_season_team_stats(self):
        return self.regular_stats.all()


class RegularSeasonTeamStats(models.Model):
    team_id = models.ForeignKey(Team, on_delete=models.PROTECT)
    season = models.ForeignKey(SeasonSummary, on_delete=models.PROTECT, related_name='regular_stats')
    conference = models.CharField()
    conference_rank = models.IntegerField()
    division = models.CharField()
    division_rank = models.IntegerField()
    win = models.IntegerField()
    lose = models.IntegerField()
    pts = models.IntegerField()
    ast = models.IntegerField()
    stl = models.IntegerField()
    blk = models.IntegerField()
    fg = models.IntegerField()
    fga = models.IntegerField()
    three = models.IntegerField()
    threea = models.IntegerField()
    ft = models.IntegerField()
    fta = models.IntegerField()
    oreb = models.IntegerField()
    dreb = models.IntegerField()
    to = models.IntegerField()
    pf = models.IntegerField()
    plusminus = models.IntegerField()
    @property
    def reb(self):
        return self.oreb + self.dreb
    @property
    def eff(self):
        return (self.pts + self.reb + self.ast + self.stl + self.blk) - ((self.fga - self.fg) + (self.fta - self.ft) + self.to)
    @property
    def team_abbreviation(self):
        return self.team_id.abbreviation
    @property
    def team_logo(self):
        return self.team_id.logo

    class Meta:
        constraints = [
            models.UniqueConstraint(fields=['team_id', 'season'], name='unique_team_id_season')
        ]
    



